File: src/csharp/Facebook.CSSLayout.Tests/CSSNodeTest.cs
/**
 * Copyright (c) 2014, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

using System;
using NUnit.Framework;

namespace Facebook.CSSLayout.Tests
{
    /**
     * Tests for {@link CSSNode}.
     */
    public class CSSNodeTest
    {

        [Test]
        public void testAddChildGetParent()
        {
            CSSNode parent = new CSSNode();
            CSSNode child = new CSSNode();

            Assert.Null(child.getParent());
            Assert.AreEqual(0, parent.getChildCount());

            parent.addChildAt(child, 0);

            Assert.AreEqual(1, parent.getChildCount());
            Assert.AreEqual(child, parent.getChildAt(0));
            Assert.AreEqual(parent, child.getParent());

            parent.removeChildAt(0);

            Assert.Null(child.getParent());
            Assert.AreEqual(0, parent.getChildCount());
        }

        [Test, ExpectedException(typeof(InvalidOperationException))]
        public void testCannotAddChildToMultipleParents()
        {
            CSSNode parent1 = new CSSNode();
            CSSNode parent2 = new CSSNode();
            CSSNode child = new CSSNode();

            parent1.addChildAt(child, 0);
            parent2.addChildAt(child, 0);
        }
    }
}

File: java/tests/com/facebook/csslayout/CSSNodeTest.java
/**
 * Copyright (c) 2014-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

package com.facebook.csslayout;

import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class CSSNodeTest {

  @Test
  public void testInit() {
    final int refCount = CSSNode.jni_CSSNodeGetInstanceCount();
    final CSSNode node = new CSSNode();
    assertEquals(refCount + 1, CSSNode.jni_CSSNodeGetInstanceCount());
  }

  @Test
  public void testMeasure() {
    final CSSNode node = new CSSNode();
    node.setMeasureFunction(new CSSNodeAPI.MeasureFunction() {
        public long measure(
            CSSNodeAPI node,
            float width,
            CSSMeasureMode widthMode,
            float height,
            CSSMeasureMode heightMode) {
          return MeasureOutput.make(100, 100);
        }
    });
    node.calculateLayout(null);
    assertEquals(100, (int) node.getLayoutWidth());
    assertEquals(100, (int) node.getLayoutHeight());
  }

  private CSSLogLevel mLogLevel;
  private String mLogMessage;

  @Test
  public void testLogger() {
    CSSNode.setLogger(new CSSLogger() {
        public void log(CSSLogLevel level, String message) {
          mLogLevel = level;
          mLogMessage = message;
        }
    });
    CSSNode.jni_CSSLog(CSSLogLevel.DEBUG.intValue(), "Hello");
    assertEquals(CSSLogLevel.DEBUG, mLogLevel);
    assertEquals("Hello", mLogMessage);
  }

  @Test
  public void testUpdateLogger() {
    CSSNode.setLogger(new CSSLogger() {
        public void log(CSSLogLevel level, String message) {}
    });
    CSSNode.setLogger(new CSSLogger() {
        public void log(CSSLogLevel level, String message) {
          mLogLevel = level;
          mLogMessage = message;
        }
    });
    CSSNode.jni_CSSLog(CSSLogLevel.VERBOSE.intValue(), "Flexbox");
    assertEquals(CSSLogLevel.VERBOSE, mLogLevel);
    assertEquals("Flexbox", mLogMessage);
  }

  @Test
  public void testCopyStyle() {
    final CSSNode node0 = new CSSNode();
    assertTrue(CSSConstants.isUndefined(node0.getMaxHeight()));

    final CSSNode node1 = new CSSNode();
    node1.setMaxHeight(100);

    node0.copyStyle(node1);
    assertEquals(100, (int) node0.getMaxHeight());
  }
}
